libGL: use headers from glvnd

gebner opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Currently, libGL depends on the dev output of mesa, which in turn depends on the mesa drivers.  This e.g. causes many programs to include all mesa drivers in their closure: #118410 

This PR changes the libGL stub to use the GL headers from libglvnd instead, thus removing the mesa build dependency.
See also #44831 and #100712 (this does not completely solve the mesa issue, since some packages still need to depend on mesa directly)

I've quickly built and tested a few packages and could not find any regressions: alacritty, blender, chromium (still references mesa though), mpv (also references mesa)
Suggestions for other packages to test are welcome (i.e., what do you expect to break).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

gebner committed:
libGL: use headers from glvnd

The glvnd library contains a copy of the OpenGL headers.  By compiling
against the glvnd headers we completely avoid a dependency on Mesa (and
LLVM).  This reduces rebuild time.  It also prevents accidents, such as
where some packages kept a reference to the mesa.drivers output and thus
had all Mesa drivers in their runtime closure.

gebner commented:
Mmmh, gtk3 no longer builds after rebasing to staging..

gebner committed:
gtk3: use libGL instead of mesa

gebner committed:
gtk4: use libGL instead of mesa

gebner committed:
chromium: add explicit mesa dependency

gebner commented:
There's quite a few packages that depend on mesa because require libgbm.  This is still a bit of a problem because `mesa.dev` still depends on the mesa drivers.  Maybe it would be good to extract libgbm to a separate derivation, or make extra outputs for gbm and gbmdev.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 118479` at d0a5ac03 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>663 packages marked as broken and skipped:</summary>
  <ul>
    <li>adobe-reader</li>
    <li>adoptopenjdk-hotspot-bin-13</li>
    <li>adoptopenjdk-hotspot-bin-14</li>
    <li>adoptopenjdk-jre-hotspot-bin-13</li>
    <li>adoptopenjdk-jre-hotspot-bin-14</li>
    <li>adoptopenjdk-jre-openj9-bin-13</li>
    <li>adoptopenjdk-jre-openj9-bin-14</li>
    <li>adoptopenjdk-openj9-bin-13</li>
    <li>adoptopenjdk-openj9-bin-14</li>
    <li>agdaPackages.iowa-stdlib</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/8ca81345510641b8aa15b22a3f2ed7d7">linuxPackages.amdgpu-pro</a> (linuxPackages_5_10.amdgpu-pro)</li>
    <li><a href="https://gist.github.com/cf835881cd9669ac15ba2f3328efc236">ocz-ssd-guru</a></li>
  </ul>
</details>
<details>
  <summary>8248 packages skipped due to time constraints:</summary>
  <ul>
    <li>AusweisApp2</li>
    <li>DisnixWebService</li>
    <li>EmptyEpsilon</li>
    <li>Fabric (python38Packages.Fabric)</li>
    <li>MIDIVisualizer</li>
    <li>OSCAR</li>
    <li>R</li>
    <li>SDL</li>
    <li>SDL2</li>
    <li>SDL2_gfx</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>82 packages built successfully:</summary>
  <ul>
    <li>adoptopenjdk-bin (adoptopenjdk-hotspot-bin-11 ,openjdk11-bootstrap)</li>
    <li>openjdk8-bootstrap (adoptopenjdk-hotspot-bin-8)</li>
    <li>asciidoctor</li>
    <li>at-spi2-atk</li>
    <li>at-spi2-core</li>
    <li>atk (gnome2.atk ,gnome3.atk)</li>
    <li>cairo (gnome2.cairo)</li>
    <li>cargo</li>
    <li>epoxy</li>
    <li>freeoffice</li>
    <li>gdk-pixbuf</li>
    <li>git</li>
    <li>glpng</li>
    <li>libsoup (gnome2.libsoup)</li>
    <li>pango (gnome2.pango)</li>
    <li>gnome3.gnome-session-ctl</li>
    <li>gobject-introspection (gnome3.gobject-introspection)</li>
    <li>gsettings-desktop-schemas (gnome3.gsettings-desktop-schemas)</li>
    <li>gtk3 (gnome3.gtk ,gtk3-x11)</li>
    <li>libgudev (gnome3.libgudev)</li>
    <li>librsvg (gnome3.librsvg)</li>
    <li>tracker (gnome3.tracker)</li>
    <li>harfbuzz</li>
    <li>hipchat</li>
    <li>iconnamingutils</li>
    <li>json-glib</li>
    <li>kisslicer</li>
    <li>lasem</li>
    <li>libGL</li>
    <li>libGLU (mesa_glu)</li>
    <li>libass</li>
    <li>libdevil</li>
    <li>libinput</li>
    <li>libmanette</li>
    <li>libnotify</li>
    <li>libqrtr-glib</li>
    <li>libva</li>
    <li>libwacom</li>
    <li>openconnect (openconnect_gnutls)</li>
    <li>python38Packages.GitPython</li>
    <li>python38Packages.aiohttp</li>
    <li>python38Packages.aiohttp-cors</li>
    <li>python38Packages.arrow</li>
    <li>python38Packages.brotlipy</li>
    <li>python38Packages.construct</li>
    <li>python38Packages.cryptography</li>
    <li>python38Packages.dateparser</li>
    <li>python38Packages.dbus-python</li>
    <li>python38Packages.jeepney</li>
    <li>python38Packages.keyring</li>
    <li>python38Packages.oauthlib</li>
    <li>python38Packages.poetry</li>
    <li>python38Packages.pycairo</li>
    <li>python38Packages.pyftpdlib</li>
    <li>python38Packages.pygobject3</li>
    <li>python38Packages.pyjwt</li>
    <li>python38Packages.pyopenssl</li>
    <li>python38Packages.requests-toolbelt (python38Packages.requests_toolbelt)</li>
    <li>python38Packages.requests_oauthlib</li>
    <li>python38Packages.secretstorage</li>
    <li>python38Packages.service-identity</li>
    <li>python38Packages.trio</li>
    <li>python38Packages.trustme</li>
    <li>python39Packages.GitPython</li>
    <li>python39Packages.aiohttp</li>
    <li>python39Packages.aiohttp-cors</li>
    <li>python39Packages.arrow</li>
    <li>python39Packages.brotlipy</li>
    <li>python39Packages.construct</li>
    <li>python39Packages.cryptography</li>
    <li>python39Packages.dateparser</li>
    <li>python39Packages.jeepney</li>
    <li>python39Packages.pycairo</li>
    <li>python39Packages.pyopenssl</li>
    <li>python39Packages.secretstorage</li>
    <li>python39Packages.service-identity</li>
    <li>python39Packages.trio</li>
    <li>python39Packages.trustme</li>
    <li>scrolls</li>
    <li>softmaker-office</li>
    <li>stoken</li>
    <li>wrapGAppsHook</li>
  </ul>
</details>
<details>
  <summary>11 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/applications/window-managers/sway/default.nix:24:5:
```
   |
24 |     (substituteAll {
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/libraries/gtk/3.x.nix:83:5:
```
   |
83 |     (fetchpatch {
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/applications/window-managers/sway/default.nix:22:5:
```
   |
22 |     ./load-configuration-from-etc.patch
   |     ^
```
</li>
    <li>warning: unused-argument

Unused argument: `gmp`.
Near pkgs/development/libraries/gtk/4.x.nix:32:3:
```
   |
32 | , gmp
   |   ^
```
</li>
    <li>warning: unused-argument

Unused argument: `x11Support`.
Near pkgs/development/libraries/gtk/3.x.nix:37:3:
```
   |
37 | , x11Support ? stdenv.isLinux
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/license-missing.md">license-missing</a>

Package is missing a license.

Near pkgs/development/libraries/mesa/stubs.nix:14:3:
```
   |
14 |   buildCommand = if stdenv.hostPlatform.isDarwin then ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/libraries/gtk/3.x.nix:81:5:
```
   |
81 |     ./patches/3.0-immodules.cache.patch
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/applications/window-managers/sway/default.nix:21:5:
```
   |
21 |     ./sway-config-no-nix-store-references.patch
   |     ^
```
</li>
    <li>warning: unused-argument

Unused argument: `gmp`.
Near pkgs/development/libraries/gtk/3.x.nix:31:3:
```
   |
31 | , gmp
   |   ^
```
</li>
    <li>warning: unused-argument

Unused argument: `x11Support`.
Near pkgs/development/libraries/gtk/4.x.nix:39:3:
```
   |
39 | , x11Support ? stdenv.isLinux
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/development/libraries/mesa/stubs.nix:14:3:
```
   |
14 |   buildCommand = if stdenv.hostPlatform.isDarwin then ''
   |   ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement.

gebner committed:
mesa: add driversdev output and test that dev doesn't depend on drivers

OpenGL applications should not depend on the Mesa drivers, since these
are supposed to be loaded at runtime from /run/opengl-driver using the
glvnd library.  Furthermore, the drivers output has a large closure
since it depends on LLVM.

The only dependency from dev to drivers came from the xatracker
pkg-config file.  This commit moves this file into a new
output called driversdev.

gebner committed:
firefox: add missing libdrm and libXdamage dependencies

gebner commented:
I've now also add a commit that breaks the dependency of mesa.dev on mesa.drivers.  Packages that use libxatracker now need to depend on mesa.driversdev.  (AFAICT this library is almost unused.)

primeos commented on pkgs/development/libraries/mesa/default.nix:
```diff
@@ -179,20 +180,20 @@ stdenv.mkDerivation {
-        substituteInPlace "$pc" --replace $out $drivers
+    # Move pkg-config files for xatracker libraries to $driversdev
+    mkdir -p $driversdev/lib/pkgconfig
+    for pc in lib/pkgconfig/xatracker.pc; do
```
A dedicated output (`driversdev`) for just `xatracker.pc` seems a bit strange to me. The name of the output is a bit misleading, the actual `xatracker` headers are in the `dev` output (I assume), and this part is difficult to review as we don't know which packages this might break (without running a full nixpkgs-review).

primeos commented on pkgs/development/libraries/mesa/stubs.nix:
```diff
@@ -61,9 +61,9 @@ stdenv.mkDerivation {
+    Version: ${libglvnd.version}
     Libs: -L${libglvnd.out}/lib -l$lib
-    Cflags: -I${mesa.dev}/include -I${libglvnd.dev}/include
+    Cflags: -I${libglvnd.dev}/include
```
This is the other part that's difficult to review. I haven't diffed `mesa.dev` and `libglvnd.dev` yet but if `mesa.dev` provides additional (or even just slightly different headers, e.g. due to older versions) this could potentially cause a lot of breakages (and due to the version change alone). Is there any official documentation for this and do we know what other distros do?

primeos reviewed (commented):
Overall the diff and idea LGTM but I have trouble to properly review this in terms of (potential) breakages.

IMO the `libGL` and `mesa` commits should also have a description that explains the rationale, impact, etc. (like "Motivation for this change" for this PR) so that that information doesn't get lost (should be in the Git history too). Depending on how long this takes to review we could also merge (some of) the other commits independently or split them into a separate PR.

primeos commented:
@vcunat what's your opinion on this?

> There's quite a few packages that depend on mesa because require libgbm. This is still a bit of a problem because mesa.dev still depends on the mesa drivers. Maybe it would be good to extract libgbm to a separate derivation, or make extra outputs for gbm and gbmdev.

I was wondering about that too. Two additional Mesa outputs certainly have a cost in terms of complexity but then again a separate `libgbm` package would be nice. But in any case we should avoid adding that to this PR as things are already complicated enough (a dedicated issue(/PR) would IMO be nice though).

gebner commented:
> the libGL and mesa commits should also have a description that explains the rationale, impact, etc. 

Ok, I'll add a rationale to the commit messages.

> Depending on how long this takes to review we could also merge (some of) the other commits independently or split them into a separate PR.

I don't think there is any subset of this PR that can be reviewed more easily.  Both the mesa and libglvnd changes affect every package that depends on opengl.  The other commits just fix the resulting issues.

> this part is difficult to review as we don't know which packages this might break (without running a full nixpkgs-review).

I'm happy for suggestions on how to better test this PR.  Rebuilding almost all of nixpkgs locally is probably not feasible.

> but if mesa.dev provides additional (or even just slightly different headers, e.g. due to older versions - wich seems very likely) this could potentially cause a lot of breakages (and due to the version change alone).

I would not expect any issues here.  If you use the nonfree nvidia drivers, you also compile against different headers (i.e., compile against mesa and link against nvidia).  That seems to work well.

Also note that we link applications against libglvnd, not mesa.  Even if the mesa headers provided more functions, applications couldn't use them because they don't exist in the libglvnd library.

> A dedicated output (driversdev) for just xatracker.pc seems a bit strange to me.

This was the easiest way to untangle the dependencies.  The `xatracker.pc` file is the only one that contains references to both the dev output (because of the headers), and to the drivers output (because of the llvm dependency).

Would you prefer to move the `xa_*.h` and `d3dadapter/*` headers to the driversdev output as well?

gebner committed:
qt514.qtwayland: fix build

gebner committed:
fcitx5-mozc: remove unused mesa dependency

FRidh commented on pkgs/development/libraries/mesa/default.nix:
```diff
@@ -210,6 +211,15 @@ stdenv.mkDerivation {
+        mkdir $out
+        echo ${self.dev} >>$out/derivations
+      '';
+      disallowedRequisites = [ llvmPackages.llvm self.drivers ];
```
In that case I think you need to be more careful with overriding, passing the overridden version in.

vcunat commented:
First, defaulting to generic headers instead of mesa's was on my wishlist long ago already, so I do like that &ndash; at least in principle, but the risks didn't seem high at a glance.  One advantage is that after that mesa will become relatively cheap to update normally (directly in nixpkgs master).  E.g. on top of your current PR a mesa update does:
```
$ ./maintainers/scripts/rebuild-amount.sh HEAD
Estimating rebuild amount by counting changed Hydra jobs.
   1386 x86_64-darwin
    725 x86_64-linux
```



> I'm happy for suggestions on how to better test this PR. Rebuilding almost all of nixpkgs locally is probably not feasible.

We can have a Hydra jobset for this branch.  I can set it up myself when you think it's a good moment.  I hope most issues would be visible as build failures.

> This was the easiest way to untangle the dependencies. The `xatracker.pc` file is the only one that contains references to both the dev output (because of the headers), and to the drivers output (because of the llvm dependency).

I still feel like missing something here.  Why do the packages retain (transitive) dependency to `.drivers` in case they don't need libxatracker?  If a package keeps _runtime_ reference to `foo.dev`, I'd typically consider it a mistake, though in some cases it may be hard to fix (/cc #118410).

gebner commented:
> If a package keeps runtime reference to foo.dev, I'd typically consider it a mistake, though in some cases it may be hard to fix (/cc #118410).

It's indeed a mistake, albeit a regular one: see #98755 for another recent case.  You also need to dev output to compile C programs that use mesa as a dependency.  It's nice if you don't have to download (unused) mesa drivers and a (second) copy of LLVM for that.

> I can set it up myself when you think it's a good moment.

I'm not sure if we should also move the headers to the driversdev output (instead of just the pkg-config file).  Aside from that, I'd say this is PR is ready.

gebner commented:
@vcunat I've rebased the PR to staging and moved the corresponding headers to the driversdev output as well.  I think it would be a good time for the hydra jobset now.

vcunat commented:
I hope it won't hit many issues that are regressions in `master..staging`; if so, we'd better "change base" for the purpose of testing.  Here's the jobset: https://hydra.nixos.org/eval/1661526

gebner committed:
wayfire: add missing mesa dependency

gebner committed:
atomEnv: add missing mesa dependency

gebner committed:
pcsxr: add missing libXxf86vm dependency

gebner committed:
byzanz: add missing libXdamage dependency

gebner commented:
Thanks for setting up the jobset!  [Compared to trunk](https://hydra.nixos.org/eval/1661526?compare=trunk&full=1#tabs-now-fail), there are indeed quite a few unrelated new failures.  These seem to be mostly python packages.

There were two packages (wayfire, and trilium-desktop) that were missing a dependency on mesa (for the EGL and libgbm, resp.).  I've fixed these plus two additional packages that were missing graphics-related dependencies.  (I didn't check, but they were probably broken before.)

gebner committed:
drawio: add missing mesa dependency

gebner committed:
cagebreak: add missing mesa dependency

gebner committed:
cage: add missing mesa headers

gebner committed:
thunderbird: add missing libdrm and libXdamage dependencies

gebner committed:
webkitgtk: add mesa dependency

gebner committed:
signal-desktop: add mesa dependency

gebner committed:
wingpanel: add mesa dependency

gebner committed:
cypress: add missing mesa dependency

gebner committed:
waybox: add missing mesa dependency

gebner committed:
kodi: add missing mesa dependency

gebner committed:
insomnia: add missing mesa dependency

gebner committed:
hikari: add mesa dependency

gebner committed:
onlyoffice-bin: add libdrm dependency

gebner committed:
mullvad-vpn: add mesa dependency

gebner committed:
wio: add missing mesa dependency

gebner committed:
streamlink-twitch-gui: add mesa dependency

gebner commented:
I've fixed all the new failures (compared to master) that seem to be mesa-related.
